File: app/src/main/java/com/metahumanz/pacilread/reader/PageSlice.java
package com.metahumanz.pacilread.reader;

public class PageSlice {
    public final int start;
    public final int end;
    public final int bodyStartInSlice;
    public final int bodyEndInSlice;
    public final CharSequence text;

    public PageSlice(int start, int end, int bodyStartInSlice, int bodyEndInSlice, CharSequence text) {
        CharSequence safeText = text == null ? "" : text;
        int textLength = safeText.length();
        this.start = Math.max(start, 0);
        this.end = Math.max(end, this.start);
        if (bodyStartInSlice < 0 || bodyEndInSlice < 0) {
            this.bodyStartInSlice = -1;
            this.bodyEndInSlice = -1;
        } else {
            this.bodyStartInSlice = Math.max(0, Math.min(bodyStartInSlice, textLength));
            this.bodyEndInSlice = Math.max(this.bodyStartInSlice, Math.min(bodyEndInSlice, textLength));
        }
        this.text = safeText;
    }

    public boolean hasBodyText() {
        return bodyStartInSlice >= 0
                && bodyEndInSlice > bodyStartInSlice
                && end > start;
    }
}

File: app/src/main/kotlin/com/metahumanz/pacilread/reader/PageSlice.kt
package com.metahumanz.pacilread.reader

class PageSlice(start: Int, end: Int, bodyStartInSlice: Int, bodyEndInSlice: Int, text: CharSequence?) {
    @JvmField val start: Int = start.coerceAtLeast(0)
    @JvmField val end: Int = end.coerceAtLeast(this.start)
    @JvmField val bodyStartInSlice: Int
    @JvmField val bodyEndInSlice: Int
    @JvmField val text: CharSequence = text ?: ""

    init {
        if (bodyStartInSlice < 0 || bodyEndInSlice < 0) {
            this.bodyStartInSlice = -1
            this.bodyEndInSlice = -1
        } else {
            this.bodyStartInSlice = bodyStartInSlice.coerceIn(0, this.text.length)
            this.bodyEndInSlice = bodyEndInSlice.coerceIn(this.bodyStartInSlice, this.text.length)
        }
    }

    fun hasBodyText(): Boolean =
        bodyStartInSlice >= 0 && bodyEndInSlice > bodyStartInSlice && end > start
}
